### PR TITLE
Fix-6911(pytest-bot): Added error from commands that are run

### PR DESCRIPTION
**Pull request by gdhameeja**
#6911 , Earlier pytest-bot would only print out the exception in cases of
failure but did not provide context on failing command and error from
command. This patch adds the errors from the command to the exception
message.
`Command` provides abstraction over the command to run and helps
in collecting errors from the first failing command only.
With this, we don't need to check `returncode` from each command that
we run, we can run all the commands and will have access to the error
from the first command that failed. This pattern was taken from Go.
Please refer: https://blog.golang.org/errors-are-values

### CODE REVIEW

File: scripts/release-on-comment.py [hunk 1]
```diff
@@ -119,19 +119,23 @@ def trigger_release(payload_path: Path, token: str) -> None:
...
 
-        check_call(
+        Command.run(
             [sys.executable, "scripts/release.py", version, "--skip-check-links"]
         )
 
         oauth_url = f"https://{token}:x-oauth-basic@github.com/{SLUG}.git"
-        check_call(["git", "push", oauth_url, f"HEAD:{release_branch}", "--force"])
+        Command.run(["git", "push", oauth_url, f"HEAD:{release_branch}", "--force"])
+        if Command.has_error():
```
Comment (gdhameeja) [inline scripts/release-on-comment.py]: Instead of checking if `completed_process.returncode != 0` after every run, we just run every command with `Command.run()`. And we just check once if any error occurred. Then we can raise an error from the first command that failed.

File: scripts/release-on-comment.py [hunk 2]
```diff
@@ -211,5 +215,32 @@ def main() -> None:
...
+    failed_command = None
+
+    @staticmethod
+    def run(command):
+        completed_process = run(command, stdout=PIPE, stderr=PIPE)
+        if completed_process.returncode != 0 and not Command.error_code:
+            Command.error = completed_process.stderr.decode()
+            # we might have numbers in command which will fail in join
+            command = [str(each) for each in command]
+            Command.failed_command = " ".join(command)
```
Comment (RonnyPfannschmidt) [inline scripts/release-on-comment.py]: Im not sure if we need it for pytest, but in pedantic correctness, shell quoting should be used

**Comment by gdhameeja** [comment]
> I would strongly prefer to have run as a factory
I don't understand the 'run as a factory' part. Could you please guide me on the same.
 
> Also i strongly recommend early bailout in any case, based on personal bias that collective errors compound unfortunately
Wouldn't I have to check for every command if there is an error in case of early bailout? And if we add more commands here later on the if error: raise error statement would grow for every command right?
Here I don't understand the 'collective errors compound' part. Could you please elaborate so I can understand this for future? Thanks.

File: scripts/release-on-comment.py (hunk 2 above)
Comment (gdhameeja) [reply scripts/release-on-comment.py]: You mean escaping special characters? Apologies, I'm new to the concept, but I can definitely research and fix it.

**Comment by RonnyPfannschmidt** [comment]
I'll compose a more detailed reply when I finish recovery from my normal cold and get back to the computer

**Comment by nicoddemus** [comment]
Hi @gdhameeja, 

First of all thanks for working on this!

Following @RonnyPfannschmidt comments/advice, there are two key points I think we should improve here:

1. We should stop immediately if any command fails; your implementation as is will continue to execute commands even if any of them fails, which will lead to undesired/unexpected results.

2. The `Command` class is essentially a global, because it has static methods which manipulate class attributes. This in general is not considered a good practice.

While it is interesting to add a `Command` abstraction here, I don't think it is actually needed, all we need is a function that:

1. Runs a command, capturing output while also forwarding it to stdout/stderr.
2. If that command fails, immediately raise an error, passing along the captured output so that we can use it to post a comment in the original issue.

So instead of `Command`, we can have a function like this:

```python
def run_capturing(command):
    completed_process = run(command, stdout=PIPE, stderr=STDOUT)
    stdout = completed_process.stdout.decode('UTF-8')
    print(stdout, end='', flush=True)
    if completed_process.returncode != 0:
        raise CalledProcessError(completed_process.returncode, command, output=stdout)
```` 

(`stderr=STDOUT` here means to redirect to `stdout` any writes to the process `stderr`; this ensure we see the output stream in the same way it appears in the terminal)


Then in `trigger_release`, we use `run_capturing` instead of `check_call`, but now we can catch `CalledProcessError` (which contain the
captured output) and use the output in the comment message:

```python
def trigger_release(payload_path: Path, token: str) -> None:
    ...
    error_contents = ''
    try:
        ...
        run_capturing(["git", "config", "user.name", "pytest bot"])
        run_capturing(["git", "config", "user.email", "pytestbot@gmail.com"])
        ...
    except CalledProcessError as e:
        error_contents = e.output
    except Exception as e:
        # fallback to an unknown error
        error_contents = f"{type(e).__name__}: {e}"

    if error_contents:
        link = f"https://github.com/{SLUG}/actions/runs/{os.environ['GITHUB_RUN_ID']}"
        issue.create_comment(
            dedent(
                f"""
            Sorry, the request to prepare release `{version}` from {base_branch} failed with:
            
            ```
            {error_contents}
            ```

            See: {link}.
            """
            )
        )
        print_and_exit(f"{Fore.RED}{e}")
```

I think that's about it. Let me know if you have any questions. 👍

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus thanks a lot for detailing this, i have nothing to add :+1:

**Comment by gdhameeja** [comment]
@nicoddemus thank you for elaborating the points. I shall modify the PR according to the details you provided. I agree a class is not needed with the approach you outlined.
> The Command class is essentially a global, because it has static methods which manipulate class > > attributes. This in general is not considered a good practice.

I did not understand this part though, I tried searching the web on the same, but didn't found anything useful. Can you please help me understand the same? Are you saying that the class is a global because it has a bunch of class variables which are modified by static methods, and no state is involved? I don't understand why is it a global.
Secondly, can you please highlight some reasons this is a bad choice and in case I wanted to go with this approach how could I make it better?
Above questions are just for the purposes of learning stuff so I can reason better about my code in the future. Thanks.

**Comment by gdhameeja** [comment]
Fixed the pull request as per suggested changes. I would still like to understand why the `Command` pattern was not appropriate just to learn. I've heard too that globals are bad, but I don't understand how `class Command` was global and what bad implications could it have being global.
Also is there any way we could fix that approach if we wanted to use the class?
Thanks.
@nicoddemus @RonnyPfannschmidt

**Comment by RonnyPfannschmidt** [comment]
@gdhameeja any class is a "global" in python, any changes to that class object are global to all users of that class

one example is that running the next command would loose the output/status of the previous command in this particular case

the common solution to use the class is to create instances of the class in order to store each command execution in its own distinct instance

**Comment by gdhameeja** [comment]
@nicoddemus @RonnyPfannschmidt Can we merge this if nothing else is required?

File: scripts/release-on-comment.py [hunk 3]
```diff
@@ -168,6 +171,22 @@ def trigger_release(payload_path: Path, token: str) -> None:
         )
         print_and_exit(f"{Fore.RED}{e}")
 
+    if error_contents:
+        link = f"https://github.com/{SLUG}/actions/runs/{os.environ['GITHUB_RUN_ID']}"
+        issue.create_comment(
+            f"""
```
Comment (bluetech) [inline scripts/release-on-comment.py]: This is missing a `dedent()` call (and has a dangling `)` below for that reason).

File: scripts/release-on-comment.py [hunk 4]
```diff
@@ -211,5 +230,13 @@ def main() -> None:
     trigger_release(Path(options.payload), options.token)
 
 
+def run_capturing(command):
```
Comment (bluetech) [inline scripts/release-on-comment.py]: I believe this wrapper is not necessary:

- Instead of doing the decoding ourselves, can pass `text=True` to `run` and then `completed_process.stdout` will be in text-mode.
- The `print` I think can be removed, because `print_and_exit()` already prints the message.
- Checking the `returncode` can be replaced by passing `check=True` to `run`.
- The output capturing can be replaced with adding `capture_output=True`  to `run`. The output will be included in the `CalledProcessError` raised by `check=True` as well.

File: scripts/release-on-comment.py [hunk 5]
```diff
@@ -168,6 +171,22 @@ def trigger_release(payload_path: Path, token: str) -> None:
         )
         print_and_exit(f"{Fore.RED}{e}")
 
+    if error_contents:
```
Comment (bluetech) [inline scripts/release-on-comment.py]: I suggest one of two things:

1. Inline this inside of the `except CallProcessError` branch, instead of putting in the end and using `error_contents` variable.

2. Or, if you want to avoid duplication, make the `except Exception as e` branch use the `error_contents` variable as well (would be `error_contents = str(e)`).

**Comment by bluetech** [comment]
@gdhameeja Thanks, I left some comments. Note that the `linting` CI job failed, due to the syntax error I pointed out.

File: scripts/release-on-comment.py (hunk 5 above)
Comment (nicoddemus) [reply scripts/release-on-comment.py]: Oh I meant to suggest 2), I probably wasn't clear enough.

File: scripts/release-on-comment.py (hunk 4 above)
Comment (nicoddemus) [reply scripts/release-on-comment.py]: > The print I think can be removed, because print_and_exit() already prints the message.

`print_and_exit()` will only print the exception, not the entire captured output, which is useful to have.

> The output capturing can be replaced with adding capture_output=True to run. The output will be included in the CalledProcessError raised by check=True as well.

Oh awesome, that I didn't know. Indeed then this wrapper is not really necessary, I agree.

Comment (nicoddemus) [reply scripts/release-on-comment.py]: We also need to then change the call of `print_and_exit` to include `stdout` and `stderr` in that case.

**Comment by gdhameeja** [comment]
Sure. I'm a little occupied, will try to squeeze this in by today/tomorrow at latest.
@bluetech Thanks

File: scripts/release-on-comment.py (hunk 3 above)
Comment (gdhameeja) [reply scripts/release-on-comment.py:177]: fixed.

File: scripts/release-on-comment.py (hunk 4 above)
Comment (gdhameeja) [reply scripts/release-on-comment.py]: Now that we've removed the wrapper. How do we use `completed_process.stdout` or `completed_process.stderr` ?
We'd have to store the last completed_process and use that?

Comment (bluetech) [reply scripts/release-on-comment.py]: Sorry, I don't really remember what `completed_process` was. Can you rephrase the question in terms of the current state of the PR?

**Comment by gdhameeja** [comment]
Can we review this again please? What else is needed here?

**Comment by gdhameeja** [comment]
> Thanks @gdhameeja and sorry about the delay!

Thank you. Do I need to make the build pass to get this merged?

**Comment by nicoddemus** [comment]
Thanks @gdhameeja!
